feat: Add SIG for Agent Development Lifecycle (ADLC) (#101)
cosai-oasis/ws4-secure-design-agentic-systems: SIGs/ADLC/adlc-scope-doc.md

> # SIG Agent Development Lifecycle (ADLC) Scope Defintion

Reviewer: Small doc nit: typo in the title: `Defintion` should be `Definition`. I would also avoid `Final Draft` while this is still in review. `Draft for Review` or just `Ready for Review` would be cleaner.

Suggested change for the title:

```suggestion
# SIG Agent Development Lifecycle (ADLC) Scope Definition
```

Suggested change for the status block:

```md
**Version:** 1.0 - Draft for Review  
**Status:** Ready for Review  
```

| **Agent Developers** | Those building and configuring agentic systems |
| **Agentic Security Framework Developers** | Those creating tooling and infrastructure for secure agent operations |
| **Security & Risk Officers (CISO/CRO)** | Those establishing IT security and business control policies |
| **Third-Party Risk Management (TPRM) Teams** | Those evaluating and procuring commercial off-the-shelf agentic solutions |
| **Enterprise Architecture Teams** | Those integrating agents into existing security ecosystems |

## 3. Methodology

**Baseline Framework:** NIST SSDF (Secure Software Development Framework) SP 800-218  
**Supplementary Framework:** OWASP SAMM (Software Assurance Maturity Model)  

#### External Resource Control Principle

Models, orchestration frameworks (LangChain, AutoGPT), APIs, data services, and registries are upstream resources separate from the agent. ADLC does not own their internal security. ADLC owns how agents consume them: provenance verification, trust establishment, capability constraints, and runtime monitoring.

#### Implement vs. Audit Distinction

Throughout the ADLC, components fall into two categories based on who owns their security:

> ADLC Implements (In Scope): Only Agent components. We directly build, configure, and enforce security controls here. Examples:

Reviewer: I think this scope line may be too tight. The ownership split makes sense, but `only Agent components` could read as if ADLC has nothing to say about tools, RAG, memory, identity, observability, or orchestration once they sit outside the agent box. Those surfaces are where a lot of agent risk shows up. Maybe say that ADLC directly implements controls in the agent entity, while also defining and auditing the controls needed for the systems the agent uses or changes. That fits the NIST AI RMF view that AI risk depends on the specific context where the system is used: https://airc.nist.gov/airmf-resources/airmf/5-sec-core/

Suggested change:

```suggestion
ADLC Implements (In Scope): Agent-owned components and the controls that govern how the agent uses or changes adjacent systems. We directly build, configure, and enforce controls in the agent entity, and we define the requirements that prerequisite systems must satisfy for the agent to use them safely. Examples:
```

#### **Orchestration Components (Execution Platform)**

Orchestration components are the execution platform for agent operations — like the Model, they're upstream resources we consume, not implement. We configure them via Agent System Instructions and enforce policies via Agent Input/Output Handling, but the platforms themselves are prerequisites.

| CoSAI Component | What It Is | How Agent Consumes It (What We Implement in Agent) | Prerequisite (What We Audit) |
|---|---|---|---|
| **External Tools and Services** | "External APIs requiring least-privilege permissions" - APIs like Salesforce, Slack, databases that agents invoke | In Agent System Instructions: Define which tools agent can access, permission manifests, capability constraints per tool<br>In Agent Input/Output Handling: Validate tool invocations before/after, sanitize I/O | Audit: That the API provider enforces authentication, rate limiting, data handling as documented. The API's internal security (Salesforce's auth, database encryption) is the provider's responsibility. |

> | **Model Memory** | "Retains context across interactions; risks include persistent attacks and unintended data transfer to third parties" - Vector stores, databases holding agent memory | In Agent System Instructions: Define memory retention policies, PII filtering rules, session isolation requirements, access control policies<br>In Agent I/O Handling: Enforce PII filtering before writes, validate memory reads | Audit: That storage infrastructure provides encryption at rest, durability, access logging, session isolation guarantees. Infrastructure team owns storage security. |

Reviewer: Memory feels under-specified here. The sensitive moment is not just where memory is stored, but when task context gets promoted into durable memory or reusable knowledge. ADLC should say when memory can be saved, what authorized the save, what evidence it is tied to, how conflicts are handled, and how a later decision can be traced back to that memory. 

OWASP Agent Memory Guard treats this as a real agent attack surface, with checks around memory reads/writes, integrity baselines, policies, snapshots, and rollback: https://owasp.org/www-project-agent-memory-guard/

Suggested replacement for this row:

```suggestion
| **Model Memory** | "Retains context across interactions; risks include persistent attacks and unintended data transfer to third parties" - Vector stores, files, databases, or platform memory holding agent context | In Agent System Instructions: Define memory retention policies, PII filtering rules, session isolation requirements, access control policies, and rules for when ephemeral context may be promoted into durable memory. In Agent I/O Handling: Enforce PII filtering before writes, validate memory reads, require provenance on durable memory writes, and reject memory updates that contain instructions, secrets, or untrusted source material that has not passed policy checks. | Audit: That storage infrastructure provides encryption at rest, durability, access logging, session isolation guarantees, write authorization, tamper evidence, rollback, and deletion support. Infrastructure team owns storage security, but ADLC owns the policy for what the agent is allowed to remember. |
```

| **Data Storage Infrastructure** | Storage for agent operational data (logs, memory, etc.) | Implement: What data we write, retention policies, purge procedures | Audit: That infrastructure encrypts at rest, has backup/recovery, access controls. Infrastructure team owns storage security. |

## 5. Differentiation from other SIGs

ADLC addresses the security of the agent entity itself, not the security of code that agents might produce.

| Dimension | Security of Agent-Produced Code (Out of Scope) | Security of the Agent Entity (ADLC Scope) |
|---|---|---|
| **Focus** | Security of the code that agents produce | Security of the agent entity itself — infrastructure, identity, lifecycle, and operational security |
| **Primary Concern** | Code artifacts, vulnerability detection in generated code | Agent behavior, trust boundaries, runtime controls, supply chain integrity |

**Note: Under refinement**

> ## 6. ADLC Phases 

Reviewer: I would add a short scoping phase before Supply Chain. Before teams pick models, tools, or memory stores, they need to write down the intended use, autonomy limits, blast radius, data classes, IAM model, human impact, compliance category, and threat model. 

Those decisions shape everything that follows. NIST AI RMF Map 1.1 covers intended purpose and deployment context, and NIST SSDF covers security requirements and threat/risk modeling during design: https://airc.nist.gov/airmf-resources/airmf/5-sec-core/ and https://csrc.nist.gov/pubs/sp/800/218/final

Suggested change:

```suggestion
## 6. Scoping and Intended Use

Before supply-chain selection begins, teams should define the agent's intended use, autonomy boundaries, blast radius, data classifications, IAM model, human impact, compliance category, and threat model. This phase determines which tools, memories, RAG sources, models, and human-approval gates are acceptable for the agent.

Primary activities:
- Define the agent's allowed goals, prohibited goals, and escalation paths.
- Map trust boundaries across users, tools, memory, RAG, identity providers, and orchestration systems.
- Classify the data the agent may read, write, remember, or disclose.
- Decide which actions require human approval before execution or before durable memory writes.

## 7. ADLC Phases 
```

If this is applied, the current `## 7. Phase Alignment Matrix` heading should also be renumbered.

> | **4. Runtime** | Active monitoring, enforcement, and continuous compliance during agent operation | Implement: Agent identity verification (presenting credentials for authentication/authorization on each action), real-time policy enforcement in Agent I/O Handling (authorization checks, tool invocation validation, content sanitization), semantic logging and decision traces, behavioral drift detection, circuit-breakers and kill switches, anomaly detection on agent behavior, human-on-the-loop escalation for high-risk decisions, agent component monitoring (skill CVE tracking, System Instructions integrity verification, agent capability health checks, behavioral drift detection for integrated skills, agent code vulnerability scanning)<br><br>Audit (Continuous): Platform validates agent identity on each request, prerequisites continue to meet requirements (infrastructure still encrypted, model serving still isolated, APIs still authenticated), framework vulnerabilities monitored, orchestration platform health, application platform security posture maintained |

Reviewer: This is the piece I most want us to name. Runtime monitoring is covered, but there is no clear control point for the post-run step where an agent reflects on a trace, saves lessons, updates memory, or changes what it will trust next time. 

I see this as different from runtime because runtime controls the current action, while reflection/knowledge consolidation controls whether anything from the current action becomes trusted future context. The security boundary is the promotion from ephemeral context into durable memory, RAG content, policy hints, or future instructions.

Hermes-style memory flows are a concrete example: before context compression, the agent gets a dedicated memory flush where it decides what facts or lessons to persist: https://vectorize.io/articles/hermes-agent-memory-explained. The attack path is that untrusted content can become trusted future context. 

OpenClaw documents the first half of that path for malicious web pages, emails, and external documents: a subagent reads attacker-controlled content and may treat hidden instructions as user commands: https://www.getopenclaw.ai/blog/openclaw-prompt-injection-guide. OpenClaw guidance also calls out MEMORY.md and memory files as prompt context and a target for memory poisoning: https://resources.learnopenclaw.ai/openclaw-security-protect-your-ai-agent-in-production/. I would add a Reflection / Knowledge Consolidation phase or subphase with controls for provenance, approval before durable memory writes, poisoning checks, retention, conflict handling, and auditability.

Suggested change:

```suggestion
| **4. Runtime** | Active monitoring, enforcement, and continuous compliance during agent operation | Implement: Agent identity verification (presenting credentials for authentication/authorization on each action), real-time policy enforcement in Agent I/O Handling (authorization checks, tool invocation validation, content sanitization), semantic logging and decision traces, behavioral drift detection, circuit-breakers and kill switches, anomaly detection on agent behavior, human-on-the-loop escalation for high-risk decisions, agent component monitoring (skill CVE tracking, System Instructions integrity verification, agent capability health checks, behavioral drift detection for integrated skills, agent code vulnerability scanning)<br><br>Audit (Continuous): Platform validates agent identity on each request, prerequisites continue to meet requirements (infrastructure still encrypted, model serving still isolated, APIs still authenticated), framework vulnerabilities monitored, orchestration platform health, application platform security posture maintained |
| **5. Reflection / Knowledge Consolidation** | Post-run review of traces, outcomes, and context before any information becomes durable memory or future agent guidance | Implement: Memory-write policy enforcement, provenance capture for saved knowledge, source attribution, conflict checks against approved policies and source-of-truth data, poisoning checks before RAG or memory updates, human approval for high-impact durable memories, retention labels, rollback markers, and audit links from future decisions back to saved knowledge.<br><br>Audit: Memory stores and orchestration platforms preserve write logs, support rollback/deletion, enforce write authorization, and expose enough metadata to distinguish ephemeral context from approved durable memory. |
```

If this is applied, the Maintenance and Decommissioning rows should be renumbered to 6 and 7.
